### PR TITLE
Linux s390x support

### PR DESCRIPTION
**Pull request by Nayana-ibm**
We have added support for Linux s390x platform in Bazel.
Could someone please review?

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: third_party/protobuf/src/google/protobuf/stubs/atomicops_internals_generic_gcc.h [hunk 1]
```diff
@@ -128,6 +128,24 @@ inline Atomic64 NoBarrier_CompareAndSwap(volatile Atomic64* ptr,
   return old_value;
 }
 
+inline Atomic64 NoBarrier_AtomicIncrement(volatile Atomic64* ptr,
```
Comment (dslomov) [inline third_party/protobuf/src/google/protobuf/stubs/atomicops_internals_generic_gcc.h]: These changes has to go to upstream protobuf.

Comment (Nayana-ibm) [inline third_party/protobuf/src/google/protobuf/stubs/atomicops_internals_generic_gcc.h]: @dslomov  The above change has been committed in protobuf master through commit id [c59473d](https://github.com/google/protobuf/commit/c59473d53eafadd126502657e5c5c33e952b67ed) 
However, it looks Bazel doesn't use protobuf master yet.
Hence, I have made changes in 
`bazel/third_party/protobuf/src/google/protobuf/stubs/atomicops_internals_generic_gcc.h`

**Comment by laszlocsomor** [comment]
@dslomov : If the changes need to be done in https://github.com/google/protobuf/, can we close this PR?

**Comment by dslomov** [comment]
@laszlocsomor, no, the changes need to happen in this PR as well, since we cannot check in ` protoc-linux-s390x.exe`

**Comment by Nayana-ibm** [comment]
@dslomov :  We are working on an approach to build bazel without uploading binaries, 
the way you have suggested with PROTOC and GRPC_JAVA_PLUGIN environment variables.

**Comment by dslomov** [comment]
@Nayana-ibm Excellent, thank you!

**Comment by Nayana-ibm** [comment]
@dslomov , 
We see that using environment variables would include multiple code changes in files like `scripts/bootstrap/compile.sh`, `third_party/protobuf/BUILD` etc. 
We can think of another approach for doing the same as mentioned below.

We can keep the same path for s390x protoc compiler binaries as done in other platforms.
For example:

`scripts/bootstrap/compile.sh:`

     
```
 if [ "${MACHINE_IS_Z}" = 'yes' ]; then
        PROTOC=${PROTOC:-third_party/protobuf/protoc-linux-s390x.exe}
       GRPC_JAVA_PLUGIN=${GRPC_JAVA_PLUGIN:-third_party/grpc/protoc-gen-grpc-java-0.15.0-linux-s390x_64.exe}
      else
        PROTOC=${PROTOC:-third_party/protobuf/protoc-linux-x86_32.exe}
        GRPC_JAVA_PLUGIN=${GRPC_JAVA_PLUGIN:-third_party/grpc/protoc-gen-grpc-java-0.15.0-linux-x86_32.exe}
      fi

```
And 

`third_party/protobuf/BUILD`

```
filegroup(
    name = "protoc",
    srcs = select({
        # protoc-windows-x86_64.exe can't deal with long path(length > 260), change to protoc-mingw.exe.
        "//third_party:windows_mingw": ["protoc-mingw.exe"],
        .
        .
        "//third_party:k8": ["protoc-linux-x86_64.exe"],
        "//third_party:piii": ["protoc-linux-x86_32.exe"],
        "//third_party:freebsd": ["protoc-linux-x86_32.exe"],
        "//third_party:s390x": ["protoc-linux-s390x.exe"],
```


This would involve user to copy the Protobuf and gRPC-java binaries at the above mentioned path, once those are compiled independently, instead of user setting environment variables. 
This would also ensure that s390x code lines remain in sync with other platforms. 

Would you allow us to make a change in Readme.md file to add a note for s390x users to build and copy the binaries in respective locations (tensorflow/third_party)?

**Comment by Nayana-ibm** [comment]
Any update on building Bazel using an approach mentioned in my last comment?

**Comment by dslomov** [comment]
Sorry I somehow missed this update :(
Yes, I think this is fine until we get a better story for protobuf compiler.

**Comment by dslomov** [comment]
@Nayana-ibm do you need any more information from us? If you want to pursue the plan from https://github.com/bazelbuild/bazel/pull/1781#issuecomment-248579432, the steps are
1. Update this PR accordingly
2. Sign the CLA 
After the PR passes the review, we can merge it.

**Comment by Nayana-ibm** [comment]
@dslomov 
Yes. We are following the plan from https://github.com/bazelbuild/bazel/pull/1781#issuecomment-248579432

We have created a new PR #1891 with CLA signed. 
We have removed binaries and can add a note in Readme.  

Hence closing this one.

**Pull request closed without merging**